### PR TITLE
Source s3: fix datetime conversion

### PR DESCRIPTION
**Pull request by davydov-d**
## What
https://github.com/airbytehq/oncall/issues/1669

## How
 - Support both datetime formats for the time of migration. Try to parse datetimes using new format, in case of failure - default to the deprecated one
 - Use `datetime.astimezone` instead of `localize`. The latter one should be used to make naive datetime objects timezone aware, whereas we need to convert datetime object to UTC tz

### CODE REVIEW

File: airbyte-integrations/connectors/source-s3/unit_tests/test_stream.py [hunk 1]
```diff
@@ -590,3 +590,14 @@ def test_schema_no_files(self, mocker):
...
+    def test_migrate_datetime_format(self):
+        current_state = {"_ab_source_file_last_modified": "2022-11-09T11:12:00+0000"}
+        latest_record = {"_ab_source_file_last_modified": "2020-11-09T11:12:00Z"}
+        stream_instance = IncrementalFileStreamS3(
+            dataset="dummy",
+            provider={"bucket": "empty"},
+            format={"filetype": "csv"},
+            path_pattern="**/prefix*.csv"
+        )
+        assert stream_instance.get_updated_state(current_state, latest_record)["_ab_source_file_last_modified"] == "2022-11-09T11:12:00Z"
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-s3/unit_tests/test_stream.py:603]: this line fails on master brunch with the same error from the oncall issue

**Review by bnchrch — APPROVED** [review]
A few notes, nothing blocking though, so Approved!

**Review by evantahler — APPROVED** [review]
Thank you!

Reminder - once this is merged, we need to un-pin the connector in Airbyte Cloud (currently pinned to v0) - https://github.com/airbytehq/airbyte-platform-internal/blob/master/cloud-config/cloud-config-seed/src/main/resources/seed/source_definitions_mask.yaml#L232-L234

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py [hunk 2]
```diff
@@ -320,10 +321,13 @@ def _get_datetime_from_stream_state(self, stream_state: Mapping[str, Any] = None
...
         if stream_state is not None and self.cursor_field in stream_state.keys():
-            state_datetime = datetime.strptime(stream_state[self.cursor_field], self.datetime_format_string)
+            try:
+                state_datetime = datetime.strptime(stream_state[self.cursor_field], self.datetime_format_string)
+            except ValueError:
+                state_datetime = datetime.strptime(stream_state[self.cursor_field], self.deprecated_datetime_format_string)
         else:
             state_datetime = datetime.strptime("1970-01-01T00:00:00Z", self.datetime_format_string)
-        return pytz.utc.localize(state_datetime)
+        return state_datetime.astimezone(pytz.utc)
```
[the remark this answers is not in the document]
Comment (davydov-d) [reply airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py:330]: localize was introduced in the PR that made syncs fail. It is not designed to do what it was expected to. It only makes naive datetime objects aware of a timezone. Instead, we need to convert it to UTC, `astimezone` is designed for that

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py [hunk 3]
```diff
@@ -51,6 +51,7 @@ def fileformatparser_map(self) -> Mapping[str, type]:
     ab_file_name_col = "_ab_source_file_url"
     airbyte_columns = [ab_additional_col, ab_last_mod_col, ab_file_name_col]
     datetime_format_string = "%Y-%m-%dT%H:%M:%SZ"
+    deprecated_datetime_format_string = "%Y-%m-%dT%H:%M:%S%z"
```
[the remark this answers is not in the document]
Comment (davydov-d) [reply airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py:56]: done

**Pull request merged**